Dots are not allowed in the path in Hostname v2

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

Setting `hostname`/`KC_HOSTNAME` with the new Hostname v2 to an URL containing a dot in the path will crash keycloak at startup with the following message:

```
ERROR: Unexpected error when starting the server in (production) mode
Error details:
java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.ensureClassInitialized(MethodHandleAccessorFactory.java:300)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.newConstructorAccessor(MethodHandleAccessorFactory.java:103)
	at java.base/jdk.internal.reflect.ReflectionFactory.newConstructorAccessor(ReflectionFactory.java:200)
	at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:70)
	at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:139)
	at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:35)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:119)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
Caused by: java.lang.IllegalArgumentException: Provided hostname is neither a plain hostname or a valid URL
	at org.keycloak.url.HostnameV2ProviderFactory.validateAndCreateUri(HostnameV2ProviderFactory.java:84)
	at org.keycloak.url.HostnameV2ProviderFactory.init(HostnameV2ProviderFactory.java:64)
	at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.<init>(QuarkusKeycloakSessionFactory.java:83)
	at org.keycloak.quarkus.runtime.KeycloakRecorder.configSessionFactory(KeycloakRecorder.java:116)
	at io.quarkus.deployment.steps.KeycloakProcessor$configureKeycloakSessionFactory343981823.deploy_4(Unknown Source)
	at io.quarkus.deployment.steps.KeycloakProcessor$configureKeycloakSessionFactory343981823.deploy(Unknown Source)
	... 26 more
```

With Keycloak 24, the setting `hostname-url`/`KC_HOSTNAME_URL` has worked in this scenario.


### Version

25.0.4

### Regression

- [X] The issue is a regression

### Expected behavior

Keycloak 25 with Hostname v2 can work with a sub path that contains dots, like Keycloak 24 did.

### Actual behavior

Keycloak crashes at startup.

### How to Reproduce?

Set `hostname`/`KC_HOSTNAME` to `https://my-example.com/auth.this/crashes`

### Anything else?

_No response_

--- a/services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java
+++ b/services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java
@@ -17,6 +17,10 @@
 
 package org.keycloak.url;
 
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Optional;
+
 import org.keycloak.Config;
 import org.keycloak.common.Profile;
 import org.keycloak.models.KeycloakSession;
@@ -24,23 +28,15 @@ import org.keycloak.provider.EnvironmentDependentProviderFactory;
 import org.keycloak.urls.HostnameProvider;
 import org.keycloak.urls.HostnameProviderFactory;
 
-import java.net.URI;
-import java.util.Optional;
-import java.util.regex.Pattern;
-
 /**
  * @author Vaclav Muzikar <vmuzikar@redhat.com>
  */
 public class HostnameV2ProviderFactory implements HostnameProviderFactory, EnvironmentDependentProviderFactory {
+    private static final String INVALID_HOSTNAME = "Provided hostname is neither a plain hostname nor a valid URL";
     private String hostname;
     private URI hostnameUrl;
     private URI adminUrl;
     private Boolean backchannelDynamic;
-
-    // Simplified regexes for hostname validations; further validations are performed when instantiating URI object
-    private static final String hostnameStringPattern = "[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?(\\.[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?)*";
-    private static final Pattern hostnamePattern = Pattern.compile("^" + hostnameStringPattern + "$");
-    private static final Pattern hostnameUrlPattern = Pattern.compile("^(http|https)://" + hostnameStringPattern + "(:\\d+)?(/[\\w-]+)*/?$");
 
     @Override
     public void init(Config.Scope config) {
@@ -57,11 +53,10 @@ public class HostnameV2ProviderFactory implements HostnameProviderFactory, Envir
 
         // Set hostname, can be either a full URL, or just hostname
         if (hostnameRaw != null) {
-            if (hostnamePattern.matcher(hostnameRaw).matches()) {
-                hostname = hostnameRaw;
-            }
-            else {
-                hostnameUrl = validateAndCreateUri(hostnameRaw, "Provided hostname is neither a plain hostname or a valid URL");
+            if (!(hostnameRaw.startsWith("http://") || hostnameRaw.startsWith("https://"))) {
+                validateAndSetHostname(hostnameRaw);
+            } else {
+                hostnameUrl = validateAndCreateUri(hostnameRaw, INVALID_HOSTNAME);
             }
         }
 
@@ -78,17 +73,36 @@ public class HostnameV2ProviderFactory implements HostnameProviderFactory, Envir
             throw new IllegalArgumentException("hostname-backchannel-dynamic must be set to false if hostname is not provided as full URL");
         }
     }
+    
+    private void validateAndSetHostname(String hostname) {
+        URI result;
+        try {
+            result = URI.create("http://"+hostname);
+        }
+        catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(INVALID_HOSTNAME, e);
+        }
+        if (result.getHost() == null || !result.getHost().equals(hostname)) {
+            throw new IllegalArgumentException(INVALID_HOSTNAME);
+        }
+        this.hostname = hostname;
+    }
 
     private URI validateAndCreateUri(String uri, String validationFailedMessage) {
-        if (!hostnameUrlPattern.matcher(uri).matches()) {
-            throw new IllegalArgumentException(validationFailedMessage);
-        }
+        URI result;
         try {
-            return URI.create(uri);
+            result = URI.create(uri);
         }
         catch (IllegalArgumentException e) {
             throw new IllegalArgumentException(validationFailedMessage, e);
         }
+        if (!Arrays.asList("http", "https").contains(result.getScheme())) {
+            throw new IllegalArgumentException(validationFailedMessage);
+        }
+        if (result.getRawUserInfo() != null || result.getRawQuery() != null || result.getRawFragment() != null) {
+            throw new IllegalArgumentException(validationFailedMessage);
+        }
+        return result;
     }
 
     @Override

--- a/services/src/test/java/org/keycloak/url/HostnameV2ProviderFactoryTest.java
+++ b/services/src/test/java/org/keycloak/url/HostnameV2ProviderFactoryTest.java
@@ -1,0 +1,62 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.url;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Test;
+import org.keycloak.utils.ScopeUtil;
+
+public class HostnameV2ProviderFactoryTest {
+
+    @Test
+    public void hostnameUrlValidationTest() throws IOException{
+        assertHostname("https://my-example.com/auth.this", true);
+        assertHostname("https://my-example.com:8080", true);
+        assertHostname("https://my-example.com/auth%20this", true);
+        assertHostname("my-example.com", true);
+        assertHostname("192.196.0.0", true);
+        assertHostname("[2001:0000:130F:0000:0000:09C0:876A:130B]", true);
+        
+        assertHostname("https://my-example.com?auth.this", false);
+        assertHostname("my-example.com/auth.this", false);
+        assertHostname("https://my-example.com:8080#fragment", false);
+        assertHostname("my-example.com:8080", false);
+        assertHostname("ldap://my-example.com/auth%20this", false);
+        assertHostname("?my-example.com", false);
+        assertHostname("192.196.0.5555", false);
+    }
+
+    private void assertHostname(String hostname, boolean valid) {
+        Map<String, String> values = new HashMap<>();
+        values.put("hostname", hostname);
+        HostnameV2ProviderFactory factory = new HostnameV2ProviderFactory();
+        try {
+            factory.init(ScopeUtil.createScope(values));
+            assertTrue(valid);
+        } catch (IllegalArgumentException e) {
+            assertFalse(valid);
+        }
+    }
+
+}

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/url/HostnameV2Test.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/url/HostnameV2Test.java
@@ -180,7 +180,7 @@ public class HostnameV2Test extends AbstractKeycloakTest {
 
     @Test
     public void testInvalidHostnameUrl() {
-        testStartupFailure("Provided hostname is neither a plain hostname or a valid URL",
+        testStartupFailure("Provided hostname is neither a plain hostname nor a valid URL",
                 "htt://127.0.0.1.nip.io", null, null, true);
     }
 

